crates/astria-config/src/lib.rs

```diff
@@ -0,0 +1,133 @@
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{
+    parse_macro_input,
+    Item,
+};
+
+#[proc_macro_attribute]
+pub fn astria_config(attr: TokenStream, item: TokenStream) -> TokenStream {
+    let item_ast = parse_macro_input!(item as Item);
+
+    let Item::Struct(ref item_struct) = item_ast else {
+        panic!("the astria_config proc macro can only be called on structs");
+    };
+
+    let struct_name = item_struct.ident.clone();
+
+    // Generates an vec of token-tree like structures of the atomicity we care about
+    let struct_fields = item_struct
+        .fields
+        .iter()
+        .map(|x| {
+            let field_ident = x.ident.clone().unwrap();
+            let field_ty = x.ty.clone();
+            let field_vis = x.vis.clone();
+
+            (field_ident, field_ty, field_vis)
+        })
+        .collect::<Vec<_>>();
+
+    // Check if log is a valid member of the config, otherwise panic
+    let true = struct_fields
+        .iter()
+        .map(|x| x.0.to_string().to_lowercase())
+        .collect::<Vec<_>>()
+        .contains(&"log".to_string())
+    else {
+        panic!("config struct does not contain a log element, which is necessary");
+    };
+
+    // Check if the macro itself has a valid attribute
+    let attr_str = attr.to_string().to_lowercase();
+    match attr_str.as_str() {
+        "conductor" | "composer" | "sequencer" | "sequencer_relayer" => {}
+        _ => panic!(
+            "Invalid attribute {attr_str}: please select a valid astria service to generate \
+             config for"
+        ),
+    };
+
+    let prefix = format!("ASTRIA_{}_", attr_str.to_uppercase());
+    let test_prefix = format!("TESTTEST_{}", prefix);
+
+    // let struct_field_ast = struct_fields
+    //     .iter()
+    //     .map(|(id, ty, vis)| {
+    //         quote! {
+    //             #vis #id: #ty
+    //         }
+    //     })
+    //     .collect::<Vec<_>>();
+
+    let code_gen_ast = quote! {
+        const ENV_PREFIX: &str = #prefix;
+
+        use serde::{Serialize, Deserialize};
+
+        // Adding serde dependencies on top of the old struct
+        #[derive(Debug, Deserialize, Serialize)]
```

Comment:
It's better to derive these explicitly, and then instead define a `trait Config: Deserialize + Serialize` that requires these bounds. At the moment this is very magic and surprising to users. How do you discover this?

Reply:
Yeah that's a great point, I'll change.